packages/contract-notifier/src/ContractNotifier.js

```diff
@@ -0,0 +1,119 @@
+      this.maxTimeTillExpiration > 259200
+        ? parseInt(this.maxTimeTillExpiration / 86400) + " days"
+        : parseInt(this.maxTimeTillExpiration / 3600) + " hours";
+    const apiUrl = this.apiEndpoint + "/global/listActive";
+    const activeContracts = await this.networker.getJson(apiUrl, { method: "post" });
+    const expiringContracts = activeContracts
+      .map((contract) => {
+        if (!contract.type || !contract.expirationTimestamp) {
+          return null;
+        }
+        let tokenName;
+        if (contract.type === "emp") {
+          tokenName = contract.tokenName;
+        } else if (contract.type === "lsp") {
+          tokenName = contract.longTokenName;
+        } else {
+          tokenName = "";
+        }
+        const expirationUtcString = new Date(contract.expirationTimestamp * 1000).toUTCString();
+        return {
+          // UMA API currently supports only Ethereum mainnet, thus chainId of 1 is hardcoded here:
+          chainId: 1,
+          address: contract.address,
+          expirationTimestamp: contract.expirationTimestamp,
+          tokenName: tokenName,
+          expirationUtcString: expirationUtcString,
+        };
+      })
+      .filter((contract) => {
+        return (
+          contract &&
+          contract.expirationTimestamp - currentTime <= this.maxTimeTillExpiration &&
+          contract.expirationTimestamp > currentTime &&
+          (!Object.keys(notifiedExpirations).includes(contract.chainId + "_" + contract.address) ||
+            contract.expirationTimestamp -
+              notifiedExpirations[contract.chainId + "_" + contract.address].notificationTimestamp >
+              this.maxTimeTillExpiration)
+        );
+      });
+
+    let mrkdwn = `Following contracts are expiring in ${expirationPeriod}:`;
+    for (let contract of expiringContracts) {
+      mrkdwn =
+        mrkdwn +
+        `\n- ${createEtherscanLinkMarkdown(contract.address, contract.chainId)}:` +
+        ` ${contract.tokenName}` +
+        ` is expiring on ${contract.expirationUtcString}`;
+    }
+    if (expiringContracts.length) {
+      this.logger.warn({
+        at: "ExpirationsNotifier",
+        message: "Expiring contracts reminder 🔔",
+        mrkdwn,
+        notificationPath: "dev-x",
+      });
+    }
+
+    // Update google datastore on notified contracts.
+    await this.updateNotifiedExpirations(expiringContracts, currentTime);
+  }
+
+  async getNotifiedExpirations() {
+    const notifiedExpirations = (await datastore.runQuery(datastore.createQuery("NotifiedExpirations")))[0];
+    return notifiedExpirations.reduce((contracts, contract) => {
+      return {
+        ...contracts,
+        [contract.chainId + "_" + contract.address]: {
+          address: contract.address,
+          chainId: contract.chainId,
+          notificationTimestamp: contract.notificationTimestamp,
+        },
+      };
+    }, {});
+  }
+
+  async updateNotifiedExpirations(notifiedContracts, currentTime) {
```

Comment:
nice internal function! how have you tested this? does it work in production right now?

Reply:
I've tested it by running application locally with env variable pointing to google service account in sandboxed project. From GCP console I manually checked that Datastore entries were updated correctly.